cli/tools/compile.rs

```diff
@@ -97,33 +101,73 @@ pub async fn compile(
   );
   validate_output_path(&output_path)?;
 
-  let mut file = std::fs::File::create(&output_path)
-    .with_context(|| format!("Opening file '{}'", output_path.display()))?;
-  let write_result = binary_writer
-    .write_bin(
-      &mut file,
-      eszip,
-      &module_specifier,
-      &compile_flags,
-      cli_options,
+  let mut temp_filename = output_path.file_name().unwrap().to_owned();
+  temp_filename.push(format!(
+    ".tmp-{}",
+    faster_hex::hex_encode(
+      &rand::thread_rng().gen::<[u8; 8]>(),
+      &mut [0u8; 16]
     )
-    .await
-    .with_context(|| format!("Writing {}", output_path.display()));
-  drop(file);
-  if let Err(err) = write_result {
-    // errored, so attempt to remove the output path
-    let _ = std::fs::remove_file(output_path);
-    return Err(err);
+    .unwrap()
+  ));
+  let temp_path = output_path.with_file_name(temp_filename);
+
+  let mut file = std::fs::File::create(&temp_path).with_context(|| {
+    format!("Opening temporary file '{}'", temp_path.display())
+  })?;
+  let write_result = if compile_flags.eszip {
+    // TODO: write npm vfs
+    let (_, _, node_modules) = binary_writer.pack_node_modules(&mut eszip)?;
+    if let Some(node_modules) = node_modules {
+      eszip.add_opaque_data(
+        "internal://node_modules".to_string(),
+        Arc::from(deno_core::serde_json::to_string(&node_modules)?.as_bytes()),
+      );
+    }
+    file.write_all(&eszip.into_bytes()).map_err(AnyError::from)
+  } else {
+    binary_writer
+      .write_bin(
+        &mut file,
+        eszip,
+        &module_specifier,
+        &compile_flags,
+        cli_options,
+      )
+      .await
   }
+  .with_context(|| format!("Writing temporary file '{}'", temp_path.display()));
+  drop(file);
 
   // set it as executable
   #[cfg(unix)]
-  {
+  let write_result = if write_result.is_ok() && !compile_flags.eszip {
     use std::os::unix::fs::PermissionsExt;
-    let perms = std::fs::Permissions::from_mode(0o777);
-    std::fs::set_permissions(output_path, perms)?;
+    let perms = std::fs::Permissions::from_mode(0o755);
+    std::fs::set_permissions(&temp_path, perms).with_context(|| {
+      format!(
+        "Setting permissions on temporary file '{}'",
+        temp_path.display()
+      )
+    })
+  } else {
+    write_result
+  };
+
+  if let Err(err) = write_result {
+    // errored, so attempt to remove the output path
+    let _ = std::fs::remove_file(temp_path);
+    return Err(err);
   }
 
+  std::fs::rename(&temp_path, &output_path).with_context(|| {
+    format!(
+      "Renaming temporary file '{}' to '{}'",
+      temp_path.display(),
+      output_path.display()
+    )
+  })?;
```

Comment:
Maybe this should be part of `write_result` so that if the rename fails it will delete the temp file?